lib/Sema/DerivedConformanceCodable.cpp

```diff
@@ -291,27 +299,61 @@ static EnumDecl *synthesizeCodingKeysEnum(TypeChecker &tc,
 
   // Each of these vars needs a case in the enum. For each var decl, if the type
   // conforms to {En,De}codable, add it to the enum.
-  auto storedProperties = type->getStoredProperties(/*skipInaccessible=*/true);
-  auto validVarCallback = [&C, &enumDC, &mutableEnumDC](VarDecl *varDecl) {
+  bool allConform = true;
+  for (auto *varDecl : target->getStoredProperties(/*skipInaccessible=*/true)) {
+    if (!varConformsToProtocol(tc, target->getDeclContext(), varDecl, proto)) {
+      tc.diagnose(varDecl->getLoc(),
+                  diag::codable_non_conforming_property_here,
+                  proto->getDeclaredType(), varDecl->getName());
+      allConform = false;
+      continue;
+    }
+
     auto *elt = new (C) EnumElementDecl(SourceLoc(), varDecl->getName(),
                                         TypeLoc(), /*HasArgumentType=*/false,
                                         SourceLoc(), nullptr, enumDC);
     elt->setImplicit();
     mutableEnumDC->addMember(elt);
-  };
+  }
 
-  if (!validateVarsConformToProtocol(tc, type->getDeclContext(),
-                                     storedProperties, proto, validVarCallback))
+  if (!allConform)
     return nullptr;
 
   // Forcibly derive conformance to CodingKey.
   tc.checkConformancesInContext(enumDC, mutableEnumDC);
 
   // Add to the type.
-  cast<IterableDeclContext>(type)->addMember(enumDecl);
+  cast<IterableDeclContext>(target)->addMember(enumDecl);
   return enumDecl;
 }
 
+/// Fetches the \c CodingKeys enum nested in \c target, potentially reaching
+/// through a typealias if the "CodingKeys" entity is a typealias.
+///
+/// This is only useful once a \c CodingKeys enum has been validated (via \c
+/// hasValidCodingKeysEnum) or synthesized (via \c synthesizeCodingKeysEnum).
+///
+/// \param C The \c ASTContext to perform the lookup in.
+///
+/// \param target The target type to look in.
+///
+/// \return A retrieved canonical \c CodingKeys enum if \c target has a valid
+/// one; \c nullptr otherwise.
+static EnumDecl *lookupEvaluatedCodingKeysEnum(ASTContext &C,
+                                               NominalTypeDecl *target) {
+  auto codingKeyDecls = target->lookupDirect(DeclName(C.Id_CodingKeys));
```

Comment:
`lookupDirect` doesn't check access, so we'll have to remember to change this later on.